spec.bs

```diff
@@ -3711,6 +3712,101 @@ RAPPOR</a> noises each coordinate of the bit vector independently, and it is par
   1. Return |histogram|.
 </div>
 
+## Get storage interest groups for owner ## {#get-storage-interest-groups-for-owner-header}
+
+*This first introductory paragraph is non-normative.*
+
+The [=get storage interest groups for owner=] algorithm returns a Web IDL array of interest groups. This allows a Protected Audience buyer within a shared storage worklet to gain insights into their users and to send Private Aggregation reports.
+
+<xmp class="idl">
+dictionary StorageInterestGroup : AuctionAdInterestGroup {
+  unsigned long long joinCount;
+  unsigned long long bidCount;
+  sequence<PreviousWin> prevWinsMs;
+  USVString joiningOrigin;
+  long long timeSinceGroupJoinedMs;
+  long long lifetimeRemainingMs;
+  long long timeSinceLastUpdateMs;
+  long long timeUntilNextUpdateMs;
+  unsigned long long estimatedSize;
+};
+</xmp>
+
+<div algorithm>
+  To <dfn>get storage interest groups for owner</dfn> given an [=origin=] |owner|:
+
+  1. Let |resultIgs| be an empty [=list=].
+  1. Let |now| be a [=moment=] equal to the [=current wall time=].
+  1. [=list/For each=] |ig| of [=user agent=]'s [=interest group set=]:
+    1. If |ig|'s [=interest group/owner=] does not equal |owner|, then [=iteration/continue=].
+    1. Let |resultIg| be an empty {{StorageInterestGroup}} dictionary.
+    1. [=map/Set=] |resultIg|["{{GenerateBidInterestGroup/owner}}"] to |ig|'s [=interest group/owner=].
+    1. [=map/Set=] |resultIg|["{{GenerateBidInterestGroup/name}}"] to |ig|'s [=interest group/name=].
+    1. [=map/Set=] |resultIg|["{{GenerateBidInterestGroup/enableBiddingSignalsPrioritization}}"] to |ig|'s [=interest group/enable bidding signals prioritization=].
+    1. If |ig|'s [=interest group/priority vector=] is not null:
+      1. [=map/Set=] |resultIg|["{{GenerateBidInterestGroup/priorityVector}}"] to |ig|'s [=interest group/priority vector=].
+    1. If |ig|'s [=interest group/seller capabilities=] is not null:
+      1. Let |resultSellerCapabilities| be an [=ordered map=].
+      1. [=map/For each=] |origin| → |originSellerCapabilities| of |ig|'s [=interest group/seller capabilities=]:
+        1. Let |serializedOrigin| be the [=origin/serialization=] of |origin|.
+        1. [=map/Set=] |resultSellerCapabilities|[|serializedOrigin|] to |originSellerCapabilities|.
+      1. [=map/Set=] |resultIg|["{{GenerateBidInterestGroup/sellerCapabilities}}"] to |resultSellerCapabilities|.

```

Comment:
Similarly, `record` and `ordered_map` (== map) should be "implicitly compatible". And the following implicitly conversions should also hold: set (== ordered set) == list == sequence

Reply:
Yes, `record` and `ordered_map` are "implicitly compatible", but the actual capabilities are stored internally as a `byte sequence`, not a `sequence<DOMString>`. So we need some conversion here. That might be kind of difficult since I don't think we actually specify how the enum is encoded into a `byte sequence` other than that it only takes a bit.

Reply:
I filed https://github.com/WICG/turtledove/issues/1316, as I feel the right approach is to define those enum fields as strings directly. Besides, in the current specification, when these fields are being set, there's no conversion either: _"If capabilityString is "interest-group-counts" or "latency-stats", then append capabilityString to sellerCapabilities."_

Thus, it looks like it's best to treat these fields as strings now? This improves readability and aligns with the direction of the proposed specification.